### ISSUE TITLE
[$100 bounty]  Remove hardcoded perfect players from population

### ISSUE THREAD

**Issue by haraschax** [haraschax/nograd#2]
To stable converge to perfect play against perfect player, there are 20% perfect players in the population. It should really learn perfect play without this. $100 bounty if someone can make a PR that learns perfect play reliably without this.

**Comment by Treasure520520** [comment]
Hi @haraschax, is this $100 bounty still available?

I can work on a focused PR that removes the hardcoded perfect-player population and keeps training self-play based, with validation showing reliable convergence to 9-move draws against perfect play. Before spending real implementation time, I want to confirm the bounty is still payable and whether CPU-only validation logs are acceptable, since the current script defaults to CUDA.

**Comment by zhaog100** [comment]
Claiming this bounty. I'll remove the hardcoded perfect players from the population.

**Comment by gwskier11-design** [comment]
I have a focused PR in progress for this. Approach: remove any need for hardcoded perfect players in the population, keep games as population self-play, add legal-move masking, randomized legal openings, draw rewards, and a rules-derived board-state fitness signal for sampled positions. I am validating that it runs on CPU/GPU and will only open the PR if the evidence is clear enough for review.

**Comment by elektropionir2** [comment]
Submitted a focused PR for this bounty path: https://github.com/haraschax/nograd/pull/5

It fixes a population-alignment bug in `Players.mate()` where `credits` and `x_games` doubled in length while the genome tensors stayed fixed. That makes the next generation's selection/accounting inconsistent and blocks reliable self-play training. The PR also removes the CUDA-only assumption so the training loop can be smoke-tested on CPU-only machines, and includes a regression test for the mating invariant.

Validation run locally on CPU:
- `python -m unittest test_training_invariants.py`
- `python batch_arena.py --device cpu --batch-size 4 --steps 2 --runs 1 --checkpoint-every 0`
- `git diff --check`

**Comment by Corvus400** [comment]
Submitted PR #7.

- https://github.com/haraschax/nograd/pull/7

It removes the bootstrap player and saved DNA files.
It replaces the old CUDA genome with a small DNA table and adds proof checks.
PR #5 has a smoke run, and PR #6 says long training and unittest were skipped.
Validation shows passing tests, 9-move draws as X and O for five seeds, 0 bad moves across 4520 states, 0 illegal moves, and a 2.810s cProfile run.

**Comment by truongsontung** [comment]
Withdrawing /try — after reviewing, the fix requires RL retraining to prove convergence without hardcoded perfect players, which is beyond a quick reliable fix and a competitor PR already exists. Moving to simpler confirmed-cash tasks.

**Comment by JoTalbot** [comment]
Octopus automation reviewed this bounty as a potential contribution target. Issue: [$100 bounty]  Remove hardcoded perfect players from population. Reward: 100.0. I am preparing a scoped implementation and will avoid duplicate work.
